tools/clang-display-context/ClangDisplayContext.cpp

```diff
@@ -0,0 +1,200 @@
+        handleDecl(Ctx.getTranslationUnitDecl());
+      }
+    }
+  }
+
+private:
+  template <typename T> 
+  bool checkIfInteresting(const T *N);
+  void handleDecl(const Decl *D);
+  void handleStmt(const Stmt *S);
+  void setEnclosingDecl(const Decl *D);
+  void writeIssueString(const std::string &S);
+  
+  ASTContext &Ctx;
+  FullSourceLoc IssueLoc;
+  const Decl *EnclosingDecl;
+
+  unsigned Line;
+  unsigned Column;
+};
+
+template <typename T>
+bool DisplayContextConsumer::checkIfInteresting(const T *Node) {
+  // Nodes are only interesting if they contain the line with the issue.
+  if (!Node)
+    return false;
+
+  FullSourceLoc L1 = Ctx.getFullLoc(Node->getLocStart());
+  FullSourceLoc L2 = 
+    Ctx.getFullLoc(Lexer::getLocForEndOfToken(Node->getLocEnd(), 0,
+                          Ctx.getSourceManager(), Ctx.getLangOpts()));
+  
+  if (!L1.isValid() || !L2.isValid())
+    return false;
+
+  if (Line == L1.getExpansionLineNumber() && 
+      Line == L2.getExpansionLineNumber())
+    return Column <= L2.getExpansionColumnNumber() &&
+           L1.getExpansionColumnNumber() <= Column;
+
+  return Line <= L2.getExpansionLineNumber() &&
+         L1.getExpansionLineNumber() <= Line;
+}
+
+void DisplayContextConsumer::handleDecl(const Decl *D) {
+  if (!D)
+    return;
+
+  if (const auto *TD = dyn_cast<TemplateDecl>(D))
+    D = TD->getTemplatedDecl();
+  
+  if (const auto *DC = dyn_cast<DeclContext>(D)) {
+    for (const Decl *D : DC->decls()) {
+      if (!checkIfInteresting<Decl>(D))
+        continue;
+
+      handleDecl(D);
+      return;
+    }
+  }
+  
+  setEnclosingDecl(D);
+  IssueLoc = Ctx.getFullLoc(D->getLocation()).getExpansionLoc();
+  
+  handleStmt(D->getBody());
+  
+  if (!IssueLoc.isValid() || IssueLoc.getExpansionLineNumber() != Line) {
+    outs() << "Line " << Line << " and column " << Column
+           << " does not specify an issue. \n";
+    return;
+  }
+
+  writeIssueString(GetIssueString(Ctx.getSourceManager(), IssueLoc, "",
+                                  "", EnclosingDecl, Ctx.getLangOpts()));
+}
+
+void DisplayContextConsumer::handleStmt(const Stmt *S) {
+  if (!checkIfInteresting<Stmt>(S))
+    return;
+
+  IssueLoc = Ctx.getFullLoc(S->getLocStart()).getExpansionLoc();
+  
+  if (const LambdaExpr *LE = dyn_cast<LambdaExpr>(S))
+    EnclosingDecl = LE->getCallOperator();
+
+  for (const Stmt *Child : S->children())
+    handleStmt(Child);
+}
+
+void DisplayContextConsumer::setEnclosingDecl(const Decl *D) {
+  // If the issue is located in a ValueDecl which is not
+  // a FunctionDecl, traverse up until the enclosing context is found.
+  // This function aims to solve issues especially with VarDecls.
+  while (dyn_cast<ValueDecl>(D) && !dyn_cast<FunctionDecl>(D)) {
+    D = dyn_cast<Decl>(D->getDeclContext());
+  }
+  EnclosingDecl = D;
+}
+
+void DisplayContextConsumer::writeIssueString(const std::string &SAIssue) {
+  // Since Location is provided as a range, the IssueString from Clang SA
```

Comment:
I prefer to provide a check name and bug type by command line arguments. But I am fine with adding this in a followup patch.